File: LEVEL6-Exceptions-Scope/ScopeExperiment.java
/**
 * Variable Scope Experiment - Java
 * Demonstrates lexical scoping and variable lifetime
 * APT1030-A Fundamentals of Programming Languages
 */

public class ScopeExperiment {
    
    // Class-level (instance) variables
    private String instanceVariable = "I am an INSTANCE variable";
    
    // Static (class) variables
    private static String staticVariable = "I am a STATIC variable";
    private static String globalVariable = "I am like a GLOBAL variable";
    
    // Method 1: Demonstrate local scope
    public void demonstrateLocalScope() {
        // Local variable - only exists inside this method
        String localVariable = "I am a LOCAL variable (inside method)";
        
        System.out.println("\n--- INSIDE METHOD ---");
        System.out.println("Local variable: " + localVariable);
        System.out.println("Instance variable: " + instanceVariable);
        System.out.println("Static variable: " + staticVariable);
        
        // localVariable cannot be accessed outside this method
    }
    
    // Method 2: Block scope (within braces)
    public void demonstrateBlockScope() {
        System.out.println("\n--- BLOCK SCOPE ---");
        
        // Variable in outer block
        int x = 10;
        System.out.println("Outer block x = " + x);
        
        if (x > 5) {
            // Variable in inner block
            int y = 20;
            System.out.println("Inner block y = " + y);
            System.out.println("Inner block can access x = " + x);
            
            // y only exists inside this if block
        }
        
        // This would cause error - y not accessible here
        // System.out.println(y);
        
        System.out.println("After block, y is gone (out of scope)");
    }
    
    // Method 3: Loop scope
    public void demonstrateLoopScope() {
        System.out.println("\n--- LOOP SCOPE ---");
        
        // Variable declared before loop (exists after loop)
        int sum = 0;
        
        for (int i = 0; i < 5; i++) {
            sum += i;
            System.out.println("  Loop iteration " + i + ", sum = " + sum);
        }
        
        System.out.println("After loop, sum = " + sum);
        // i is NOT accessible here (scope limited to loop)
        // System.out.println(i); // ERROR!
    }
    
    // Method 4: Parameter scope
    public void demonstrateParameterScope(String parameter) {
        System.out.println("\n--- PARAMETER SCOPE ---");
        System.out.println("Parameter value: " + parameter);
        
        // Parameter acts like a local variable
        parameter = "Modified inside method";
        System.out.println("Modified parameter: " + parameter);
    }
    
    // Method 5: Shadowing (local variable hides class variable)
    public void demonstrateShadowing() {
        System.out.println("\n--- SHADOWING ---");
        
        // Instance variable
        String instanceVariable = "LOCAL version (shadows instance)";
        
        System.out.println("Local variable: " + instanceVariable);
        System.out.println("Instance variable (use 'this'): " + this.instanceVariable);
    }
    
    // Method 6: Static vs Instance context
    public static void demonstrateStaticContext() {
        System.out.println("\n--- STATIC CONTEXT ---");
        
        // Can access static variables directly
        System.out.println("Static variable: " + staticVariable);
        
        // Cannot access instance variables directly in static method
        // System.out.println(instanceVariable); // ERROR!
        
        // Must create object to access instance variables
        ScopeExperiment obj = new ScopeExperiment();
        System.out.println("Instance variable via object: " + obj.instanceVariable);
    }
    
    // Experiment 1: Access local variable outside method
    public static void experiment1() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("EXPERIMENT 1: Local Variable Outside Method");
        System.out.println("=".repeat(60));
        
        class LocalDemo {
            public String createLocal() {
                String localVar = "I am local to createLocal()";
                System.out.println("Inside method: " + localVar);
                return localVar;
            }
        }
        
        LocalDemo demo = new LocalDemo();
        String result = demo.createLocal();
        System.out.println("Returned value: " + result);
        
        // localVar is NOT accessible here
        System.out.println("\n❌ 'localVar' is not accessible outside createLocal()");
        System.out.println("   Java uses lexical (static) scope");
    }
    
    // Experiment 2: Variable lifetime
    public static void experiment2() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("EXPERIMENT 2: Variable Lifetime");
        System.out.println("=".repeat(60));
        
        // Local variable - created on stack
        int stackVar = 100;
        System.out.println("Stack variable (local): " + stackVar);
        
        // Object - created on heap
        StringBuilder heapVar = new StringBuilder("Heap object");
        System.out.println("Heap variable (object): " + heapVar);
        
        System.out.println("\nVariable Lifetime:");
        System.out.println("  • Stack variables: Exist only within their scope (method/block)");
        System.out.println("  • Heap objects: Exist as long as referenced");
        System.out.println("  • Static variables: Exist for entire program lifetime");
    }
    
    // Experiment 3: Class scope
    public static void experiment3() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("EXPERIMENT 3: Class Scope");
        System.out.println("=".repeat(60));
        
        ScopeExperiment obj1 = new ScopeExperiment();
        ScopeExperiment obj2 = new ScopeExperiment();
        
        // Each object has its own instance variable
        obj1.instanceVariable = "Object 1's value";
        obj2.instanceVariable = "Object 2's value";
        
        System.out.println("obj1.instanceVariable = " + obj1.instanceVariable);
        System.out.println("obj2.instanceVariable = " + obj2.instanceVariable);
        
        // Static variable is shared across all instances
        staticVariable = "Shared across ALL objects";
        System.out.println("\nStatic variable (shared): " + staticVariable);
        System.out.println("obj1 sees: " + obj1.staticVariable);
        System.out.println("obj2 sees: " + obj2.staticVariable);
    }
    
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("   JAVA SCOPE EXPERIMENTS");
        System.out.println("=".repeat(60));
        
        ScopeExperiment demo = new ScopeExperiment();
        
        // Run experiments
        demo.demonstrateLocalScope();
        demo.demonstrateBlockScope();
        demo.demonstrateLoopScope();
        demo.demonstrateParameterScope("Test Parameter");
        demo.demonstrateShadowing();
        demonstrateStaticContext();
        
        experiment1();
        experiment2();
        experiment3();
        
        // Summary
        System.out.println("\n" + "=".repeat(60));
        System.out.println("SCOPE SUMMARY");
        System.out.println("=".repeat(60));
        System.out.println("""
        
        Java Scope Types:
        
        1. Class/Static Scope:
           • Static variables (exist for program lifetime)
           • Accessible from static methods
        
        2. Instance Scope:
           • Non-static variables (tied to object)
           • Accessible via 'this' or object reference
        
        3. Method/Local Scope:
           • Variables declared inside methods
           • Exist only during method execution
           • Stored on stack
        
        4. Block Scope:
           • Variables declared in { } blocks
           • Exist only within that block
        
        5. Parameter Scope:
           • Method parameters
           • Act like local variables
        
        Key Points:
        ✓ Java uses lexical (static) scope
        ✓ Local variables cannot be accessed outside their block
        ✓ Use 'this' to access instance variables when shadowed
        ✓ Variables have different lifetimes (stack vs heap)
        """);
    }
}

File: LEVEL6-Exceptions-Scope/scope_experiment.py
#!/usr/bin/env python3
"""
Variable Scope Experiment - Python
Demonstrates lexical scoping and variable lifetime
APT1030-A Fundamentals of Programming Languages
"""

# Global variable (module scope)
global_variable = "I am a GLOBAL variable"
another_global = "This is also global"

def function_scope_demo():
    """Demonstrate function scope (local variables)"""
    
    # Local variable - only exists inside this function
    local_variable = "I am a LOCAL variable (inside function)"
    
    print("\n--- INSIDE FUNCTION ---")
    print(f"Inside function - Local variable: {local_variable}")
    print(f"Inside function - Global variable: {global_variable}")
    
    # Modifying global variable (requires global keyword)
    # global global_variable  # Uncomment to modify
    # global_variable = "Modified inside function"
    
    return local_variable

def nested_scope_demo():
    """Demonstrate nested scope (closure)"""
    
    outer_variable = "I am in OUTER function"
    
    def inner_function():
        # Can access outer variable (closure)
        inner_variable = "I am in INNER function"
        print(f"  Inner function sees: {outer_variable}")
        print(f"  Inner function sees: {inner_variable}")
        return inner_variable
    
    print("\n--- NESTED SCOPE ---")
    result = inner_function()
    print(f"Outer function got: {result}")
    
    # This would cause error - inner_variable not accessible here
    # print(inner_variable)
    
    return outer_variable

def scope_experiment_1():
    """Experiment 1: Access local variable outside function"""
    print("\n" + "=" * 60)
    print("EXPERIMENT 1: Local Variable Outside Function")
    print("=" * 60)
    
    def create_local():
        x = "I am local to create_local()"
        print(f"Inside function: {x}")
        return x
    
    result = create_local()
    print(f"Returned value: {result}")
    
    # TRYING TO ACCESS LOCAL VARIABLE DIRECTLY
    print("\nAttempting to access 'x' directly...")
    try:
        print(x)  # This will cause NameError
    except NameError as e:
        print(f"❌ ERROR: {e}")
        print("   'x' only exists inside create_local() function!")
    
    print("\n✅ CONCLUSION: Local variables are NOT accessible outside their function")

def scope_experiment_2():
    """Experiment 2: Global vs Local variables"""
    print("\n" + "=" * 60)
    print("EXPERIMENT 2: Global vs Local Variables")
    print("=" * 60)
    
    # Global variable
    value = "GLOBAL"
    print(f"Initial global value: {value}")
    
    def change_local():
        # This creates a LOCAL variable, doesn't modify global
        value = "LOCAL"
        print(f"Inside function (local): {value}")
    
    def change_global():
        # This modifies the GLOBAL variable
        global value
        value = "MODIFIED GLOBAL"
        print(f"Inside function (global): {value}")
    
    change_local()
    print(f"After change_local(): {value}")
    
    change_global()
    print(f"After change_global(): {value}")

def scope_experiment_3():
    """Experiment 3: Variable lifetime"""
    print("\n" + "=" * 60)
    print("EXPERIMENT 3: Variable Lifetime")
    print("=" * 60)
    
    def create_counter():
        count = 0  # Created when function is called
        
        def increment():
            nonlocal count  # Use nonlocal for nested functions
            count += 1
            return count
        
        return increment
    
    counter = create_counter()
    print(f"Counter 1: {counter()}")  # count is preserved between calls
    print(f"Counter 2: {counter()}")
    print(f"Counter 3: {counter()}")
    
    # The 'count' variable persists due to closure
    print("\n✅ The 'count' variable continues to exist (closure)")

def scope_experiment_4():
    """Experiment 4: LEGB Rule (Local, Enclosing, Global, Built-in)"""
    print("\n" + "=" * 60)
    print("EXPERIMENT 4: LEGB Scope Resolution")
    print("=" * 60)
    
    # Global variable
    name = "Global"
    
    def outer():
        # Enclosing variable
        name = "Enclosing"
        
        def inner():
            # Local variable
            name = "Local"
            print(f"  Local scope: {name}")
            return name
        
        print(f"Enclosing scope: {name}")
        inner()
        return name
    
    print(f"Global scope: {name}")
    outer()
    
    # Built-in scope example
    print(f"\nBuilt-in scope example: {len('Hello')}")  # len() is built-in

def main():
    """Main function running all scope experiments"""
    
    print("=" * 60)
    print("   PYTHON SCOPE EXPERIMENTS")
    print("=" * 60)
    
    # Run experiments
    scope_experiment_1()
    scope_experiment_2()
    scope_experiment_3()
    scope_experiment_4()
    
    # Summary
    print("\n" + "=" * 60)
    print("SCOPE SUMMARY")
    print("=" * 60)
    print("""
    Python Scope Rules (LEGB):
    
    L - Local:     Variables defined inside a function
    E - Enclosing: Variables in enclosing functions (closures)
    G - Global:    Variables defined at module level
    B - Built-in:  Python's built-in names (len, print, etc.)
    
    Key Points:
    ✓ Local variables cannot be accessed outside their function
    ✓ Use 'global' keyword to modify global variables
    ✓ Use 'nonlocal' keyword to modify enclosing variables
    ✓ Variables can have different lifetimes (stack vs closure)
    """)

if __name__ == "__main__":
    main()